File: client/src/lib/poker-logic/hand.ts
import { createInitalDeck } from "./deck";
import { rateHand } from "./best-hand-calculator";
import type { InitialPlayer, Player, HandState } from "./model";
import { botGameState, currentBotGameState } from "../ui-logic/state";
import { logEvent } from "../analytics/analytics";

export function createInitalHandState(
  initialPlayers: InitialPlayer[],
  pot: number
): HandState {
  const smallBlind = 1;
  const bigBlind = 2;
  const deck = createInitalDeck();
  const seats = initialPlayers.map((player, index): Player => {
    let stack = player.stack;
    let lastAction: "None" | "Big Blind" | "Small Blind" = "None";
    let currentRaise = 0;
    if (index === 0) {
      stack -= smallBlind;
      currentRaise = smallBlind;
      lastAction = "Small Blind";
    }
    if (index === 1) {
      stack -= bigBlind;
      lastAction = "Big Blind";
      currentRaise = bigBlind;
    }
    return {
      cards: [deck.pop(), deck.pop()],
      stack: stack,
      isCurrentPlayer: player.isCurrentPlayer,
      out: false,
      lastAction: lastAction,
      currentRaise: currentRaise,
    };
  });
  const communityCards = [
    deck.pop(),
    deck.pop(),
    deck.pop(),
    deck.pop(),
    deck.pop(),
  ];
  return {
    seats: seats,
    round: "Blinds",
    currentAction: {
      seatInTurn: (1 + 1) % seats.length, // Left of big blind
      minRaise: bigBlind,
      lastSeatToRaise: -1,
    },
    communityCards: communityCards,
    pot: pot + smallBlind + bigBlind,
    deck: deck,
    finished: false,
    winners: [],
  };
}

export function prepareNextHand(pokerState: HandState): HandState {
  if (pokerState.seats.filter((it) => it.stack === 0).length === 1) {
    return pokerState;
  }
  const seats: InitialPlayer[] = pokerState.seats.map((seat) => {
    return {
      isCurrentPlayer: seat.isCurrentPlayer,
      stack: seat.stack,
    };
  });

  const lastSeat = seats.pop();
  seats.unshift(lastSeat);

  return createInitalHandState(seats, pokerState.pot);
}

function handlePayouts(pokerState: HandState): HandState {
  const winnings = Math.floor(pokerState.pot / pokerState.winners.length);
  const excess = pokerState.pot - winnings * pokerState.winners.length;
  pokerState.winners.forEach((seat) => {
    pokerState.seats[seat].stack += winnings;
  });
  pokerState.pot = excess;
  return pokerState;
}

function finishHand(pokerState: HandState): HandState {
  pokerState = handlePayouts(pokerState);
  pokerState.finished = true;
  const gameFinished =
    pokerState.seats.filter((it) => it.stack === 0).length === 1;
  if (gameFinished) {
    let newRank = Math.max(0, -5);
    if (
      pokerState.winners.filter(
        (winnerseat) => pokerState.seats[winnerseat].isCurrentPlayer
      ).length > 0
    ) {
      newRank = currentBotGameState.currentScore + 5;
    }
    currentBotGameState.currentScore = newRank;
    botGameState.set(currentBotGameState);
  }
  logEvent("hand-finished", {
    gameFinished,
  });
  return pokerState;
}

function finishTurn(pokerState: HandState): HandState {
  console.debug({
    message: "Finishing turn.",
  });
  logEvent("turn-finished");
  const playersIn = pokerState.seats.filter((it) => !it.out);
  if (playersIn.length === 1) {
    pokerState.winners = [pokerState.seats.findIndex((it) => !it.out)];
    return finishHand(pokerState);
  }

  let seatInTurn = pokerState.currentAction.seatInTurn;
  seatInTurn = (seatInTurn + 1) % pokerState.seats.length;
  while (pokerState.seats[seatInTurn].out === true) {
    seatInTurn = (seatInTurn + 1) % pokerState.seats.length;
  }
  if (seatInTurn === pokerState.currentAction.lastSeatToRaise) {
    pokerState = finishRound(pokerState);
  } else if (pokerState.currentAction.lastSeatToRaise === -1) {
    console.debug({
      message: "Updating last seat to raise",
      previous: pokerState.currentAction.lastSeatToRaise,
      new: pokerState.currentAction.seatInTurn,
    });
    pokerState.currentAction.lastSeatToRaise =
      pokerState.currentAction.seatInTurn;
  }
  pokerState.currentAction.seatInTurn = seatInTurn;

  return pokerState;
}

export function finishTurnForPlayer(
  seat: number,
  pokerState: HandState
): HandState {
  if (pokerState.currentAction.seatInTurn !== seat) {
    console.debug({
      message: "Attempted to finish turn for wrong user.",
      seat: seat,
      seatInTurn: pokerState.currentAction.seatInTurn,
    });
    return pokerState;
  }
  return finishTurn(pokerState);
}

function finishRound(pokerState: HandState): HandState {
  console.debug({ message: "Round finished", pokerState });
  logEvent("round-finished");
  pokerState.currentAction = {
    seatInTurn: 0,
    minRaise: 0,
    lastSeatToRaise: -1,
  };
  pokerState.seats = pokerState.seats.map((seat) => {
    return {
      cards: seat.cards,
      stack: seat.stack,
      out: seat.out,
      isCurrentPlayer: seat.isCurrentPlayer,
      lastAction: "None",
      currentRaise: 0,
    };
  });

  const round = pokerState.round;
  if (round === "Blinds") {
    pokerState.round = "Flop";
  }
  if (round === "Flop") {
    pokerState.round = "River";
  }
  if (round === "River") {
    pokerState.round = "Turn";
  }
  const everyoneAllIn = pokerState.seats.every(
    (seat) => seat.out || seat.stack == 0
  );
  if (round === "Turn" || everyoneAllIn) {
    pokerState.seats = pokerState.seats.map((seat) => {
      seat.handStrength = rateHand([
        ...seat.cards,
        ...pokerState.communityCards,
      ]).handStrength;
      return seat;
    });
    pokerState.winners = calculateWinners(pokerState);
    return finishHand(pokerState);
  }
  return pokerState;
}

function calculateWinners(pokerState: HandState): number[] {
  const handRatings = pokerState.seats.map((seat) => {
    const score = rateHand([...seat.cards, ...pokerState.communityCards]).score;
    console.debug({
      cards: seat.cards,
      score: score,
    });
    return score;
  });

  const winningRating = [...handRatings].sort((a, b) => b - a)[0];

  const winners = handRatings
    .map((it, index) => {
      if (it === winningRating) return index;
    })
    .filter((it) => it !== undefined);

  return winners;
}

File: pvp-server/src/logic/hand.go
package logic

import (
	"log"
	"math"
	"math/rand"
)

type InitalPlayer struct {
	PlayerId int
	Stack    int
}

func CreateInitialHandState(players []InitalPlayer, smallBlindAmount int, bigBlindAmount int, remainingPot int) HandState {
	var hand = HandState{}
	deck := createInitialDeck()

	hand.Seats = []Seat{}
	for index, player := range players {
		card1, d := pop(deck)
		card2, d := pop(d)
		deck = d

		lastAction := "None"
		currentRaise := 0
		stack := player.Stack
		if index == 0 {
			currentRaise = smallBlindAmount
			stack -= smallBlindAmount
			lastAction = "Small Blind"
		}
		if index == 1 {
			currentRaise = bigBlindAmount
			stack -= bigBlindAmount
			lastAction = "Big Blind"
		}
		hand.Seats = append(hand.Seats, Seat{PlayerId: player.PlayerId, Cards: []Card{card1, card2}, Stack: stack, Out: false, LastAction: lastAction, CurrentRaise: currentRaise, HandStrength: "None"})
		hand.CommunityCards = []Card{}
		for i := 0; i < 5; i++ {
			card, newDeck := pop(deck)
			deck = newDeck
			hand.CommunityCards = append(hand.CommunityCards, card)
		}
		hand.Round = "Blinds"
		hand.CurrentAction = CurrentAction{
			SeatInTurn: 2 % len(hand.Seats), // Left of BB
			MinRaise:   bigBlindAmount, LastSeatToRaise: -1}
		hand.SmallBlindAmount = smallBlindAmount
		hand.BigBlindAmount = bigBlindAmount
		hand.Pot = smallBlindAmount + bigBlindAmount + remainingPot
		hand.Deck = deck
		hand.Finished = false
		hand.Winners = []int{}
	}
	return hand
}

func createInitialDeck() []Card {
	deck := []Card{}
	suits := []string{"hearts", "spades", "clubs", "diamonds"}
	for _, suit := range suits {
		for rank := 1; rank <= 13; rank++ {
			deck = append(deck, Card{Rank: rank, Suit: suit})
		}
	}

	rand.Shuffle(len(deck), func(i, j int) { deck[i], deck[j] = deck[j], deck[i] })
	return deck
}

func pop(cards []Card) (Card, []Card) {
	card := cards[0]
	return card, cards[1:]
}

func FinishTurnForSeat(seat int, hand HandState) HandState {
	if hand.CurrentAction.SeatInTurn != seat {
		log.Printf("Attempted to finish turn from wrong seat. seat[%v] seatInTurn[%v]\n", seat, hand.CurrentAction.SeatInTurn)
		return hand
	}
	log.Printf("Finishing turn\n")
	playersIn := []Seat{}
	playersInIndices := []int{}
	for index, player := range hand.Seats {
		if !player.Out {
			playersIn = append(playersIn, player)
			playersInIndices = append(playersInIndices, index)
		}
	}
	if len(playersIn) == 1 {
		hand.Winners = playersInIndices
		return finishHand(hand)
	}

	seatInTurn := hand.CurrentAction.SeatInTurn
	seatInTurn = (seatInTurn + 1) % len(hand.Seats)

	for hand.Seats[seatInTurn].Out == true {
		seatInTurn = (seatInTurn + 1) % len(hand.Seats)
	}

	if seatInTurn == hand.CurrentAction.LastSeatToRaise {
		hand.CurrentAction.SeatInTurn = seatInTurn
		return finishRound(hand)
	}
	if hand.CurrentAction.LastSeatToRaise == -1 {
		hand.CurrentAction.LastSeatToRaise = hand.CurrentAction.SeatInTurn
	}
	hand.CurrentAction.SeatInTurn = seatInTurn
	return hand
}

func finishRound(hand HandState) HandState {
	hand.CurrentAction = CurrentAction{SeatInTurn: 0, MinRaise: 0, LastSeatToRaise: -1}

	everyoneAllIn := true
	seats := []Seat{}
	for _, seat := range hand.Seats {
		seats = append(seats, Seat{
			PlayerId:     seat.PlayerId,
			Cards:        seat.Cards,
			Stack:        seat.Stack,
			Out:          seat.Out,
			LastAction:   "None",
			CurrentRaise: 0,
		})
		log.Printf("seat[%v]", seat)
		if seat.Out == false && seat.Stack > 0 {
			everyoneAllIn = false
		}
	}

	hand.Seats = seats
	round := hand.Round
	if round == "Blinds" {
		hand.Round = "Flop"
	}
	if round == "Flop" {
		hand.Round = "Turn"
	}
	if round == "Turn" {
		hand.Round = "River"
	}

	if round == "River" || everyoneAllIn {
		for _, seat := range hand.Seats {
			cards := append(seat.Cards, hand.CommunityCards...)
			seat.HandStrength = RateHand(cards).HandStrength
		}
		hand.Winners = calculateWinners(hand)
		return finishHand(hand)
	}
	return hand
}

func finishHand(hand HandState) HandState {
	hand = handlePayouts(hand)
	hand.Finished = true
	return hand
}

func calculateWinners(hand HandState) []int {
	handRatings := []float64{}
	for _, seat := range hand.Seats {
		cards := append(seat.Cards, hand.CommunityCards...)
		score := RateHand(cards).Score
		handRatings = append(handRatings, score)
	}

	winningRating := 0.0
	for _, rating := range handRatings {
		if rating > winningRating {
			winningRating = rating
		}
	}
	winners := []int{}
	for index, rating := range handRatings {
		if rating == winningRating {
			winners = append(winners, index)
		}
	}

	return winners
}

func handlePayouts(hand HandState) HandState {
	winnings := int(math.Floor(float64(hand.Pot) / float64(len(hand.Winners))))
	excess := hand.Pot - winnings*len(hand.Winners)

	for _, winner := range hand.Winners {
		hand.Seats[winner].Stack += winnings
	}
	hand.Pot = excess
	return hand
}
